src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -1670,21 +1673,25 @@ private void CheckModelAndSyntaxNodeToSpeculate(CSharpSyntaxNode syntax)
                         var reduced = extensionMethod.ReduceExtensionMethod(containingType, Compilation);
                         if ((object)reduced != null)
                         {
-                            results.Add(reduced.GetPublicSymbol());
+                            builder.Add(reduced.GetPublicSymbol());
                         }
                     }
                 }
 
                 lookupResult.Free();
             }
 
-            ImmutableArray<ISymbol> sealedResults = results.ToImmutableAndFree();
-            return name == null
-                ? FilterNotReferencable(sealedResults)
-                : sealedResults;
+            var result = name == null
+                ? FilterNotReferenceable(builder)
+                : builder.ToImmutable();
+
+            builder.Clear();
```

Comment:
Let's supply a policy to the pool that clears on return so we don't have to remember all the time.

Reply:
What I really wanted was a PooledSegmentedList class similar to PooledHashSet/PooledDictionary. I wanted that so I wouldn't have to define the pool myself and that the pool could be shared with other code. I'm sure there is a good reason why Roslyn doesn't have that and requires everyone to create their pool, perhaps Sam knows?